tsl/src/continuous_aggs/invalidation.c

```diff
@@ -458,6 +462,53 @@ invalidation_entry_reset(Invalidation *entry)
 	MemSet(entry, 0, sizeof(Invalidation));
 }
 
+/*
+ * Expand an invalidation to bucket boundaries.
+ *
+ * Since a refresh always materializes full buckets, we can safely expand an
+ * invalidation to bucket boundaries and in the process merge a lot more
+ * invalidations.
+ */
+static void
+invalidation_expand_to_bucket_boundaries(Invalidation *inv, Oid timetype, int64 bucket_width)
+{
+	int64 min_bucket_start;
+	int64 max_bucket_end;
+
+	/* If the min value falls on the exact start of a full bucket we are
+	 * good. Otherwise, we need to move to the next full bucket. */
+	min_bucket_start =
+		ts_time_saturating_add(ts_time_get_min(timetype), bucket_width - 1, timetype);
+	min_bucket_start = ts_time_bucket_by_type(bucket_width, min_bucket_start, timetype);
+
+	/* To get the end of the last full bucket, we need to bucket the max value
+	 * and move one step down to the end value of the previous
+	 * bucket. Remember that invalidations are inclusive, so the "greatest"
+	 * value should be the last value of the last full bucket. */
+	max_bucket_end = ts_time_bucket_by_type(bucket_width, ts_time_get_max(timetype), timetype);
+	max_bucket_end = ts_time_saturating_sub(max_bucket_end, 1, timetype);
+
+	if (inv->lowest_modified_value < min_bucket_start)
+		inv->lowest_modified_value = PG_INT64_MIN;
+	else if (inv->lowest_modified_value > max_bucket_end)
+		inv->lowest_modified_value = PG_INT64_MAX;
```

Comment:
Isn't this a problem if `PG_INT16_MIN` or `PG_INT64_MAX` happen to be in the middle of a bucket?